src/components/common/custom-bottom-bar/Tab.tsx

```diff
@@ -0,0 +1,97 @@
+import { memo, useMemo } from 'react';
+import { Pressable } from 'react-native';
+import Animated, {
+  interpolate,
+  useAnimatedStyle,
+  useSharedValue,
+  withSpring,
+  withTiming,
+} from 'react-native-reanimated';
+
+import { TabList } from '~app/(tabs)/_layout';
+import { Icon, Stack, Text } from '~components/uikit';
+import { styled } from '~styles';
+
+const ANIMATION_DURATION = 350;
+
+type TabBarButtonProps = {
+  onPress: () => void;
+  onLongPress: () => void;
+  isFocused: boolean;
+  label: string;
+  tab: TabList[number];
+};
+
+const TabBarButton = ({
+  isFocused,
+  label,
+  tab,
+  ...pressableProps
+}: TabBarButtonProps) => {
+  const iconScale = useSharedValue(isFocused ? 1 : 0);
+  const labelOpacity = useSharedValue(isFocused ? 1 : 0);
+
+  const updateFocusAnimation = (focused: boolean) => {
+    iconScale.value = withSpring(focused ? 1 : 0, {
+      duration: ANIMATION_DURATION,
+    });
+    labelOpacity.value = withSpring(focused ? 1 : 0, {
+      duration: ANIMATION_DURATION,
+    });
+  };
+
+  useMemo(() => updateFocusAnimation(isFocused), [isFocused]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const animatedIconStyle = useAnimatedStyle(() => ({
+    transform: [{ scale: interpolate(iconScale.value, [0, 1], [1, 1.2]) }],
+    top: interpolate(iconScale.value, [0, 1], [0, 8]),
+  }));
+
+  const animatedTextStyle = useAnimatedStyle(() => ({
+    opacity: interpolate(labelOpacity.value, [0, 1], [1, 0]),
+  }));
+
+  return (
+    <Wrapper
+      {...pressableProps}
+      onPressIn={() => (iconScale.value = withTiming(0.8, { duration: 150 }))}
+      onPressOut={() => (iconScale.value = withTiming(1, { duration: 150 }))}
+      accessibilityRole="button"
+      accessibilityLabel={label}
+    >
+      <Stack axis="y" align="center" spacing="xs">
+        <Animated.View style={animatedIconStyle}>
+          <Icon
+            name={isFocused ? tab.iconFilled : tab.iconOutlined}
+            color={isFocused ? 'primary' : 'neutral2'}
+          />
+        </Animated.View>
+        <LabelWrapper style={animatedTextStyle}>
+          <StyledLabel variant="bodyExtraSmall" isFocused={isFocused}>
+            {label}
+          </StyledLabel>
+        </LabelWrapper>
+      </Stack>
+    </Wrapper>
+  );
+};
+
+const Wrapper = styled(Pressable, {
+  flex: 1,
+});
+
+const LabelWrapper = Animated.createAnimatedComponent(styled('View', {}));
+
+const StyledLabel = styled(Text, {
+  variants: {
+    isFocused: {
+      true: { color: '$primary' },
+      false: { color: '$textMuted' },
+    },
+  },
+});
+
+export const TabButton = memo(
```

Comment:
Curious: why do we need to memo this?

The current equality function might cause issues later on if the tab bar will have things like notification badges etc that update dynamically.

Reply:
Without the memo all the tabs are re-rendered whenever you change tabs, while with this it only re-renders the two concerned by the transition.

I agree it is quite a small optimization and it could cause issues if the tab gets more custom props.